app/models/spree/invoice.rb

```diff
@@ -0,0 +1,128 @@
+class Spree::Invoice < ActiveRecord::Base
+  belongs_to :invoiceable, polymorphic: true
+  validates :invoiceable, presence: true
+
+  before_save :set_date
+  before_save :set_number, if: :use_sequential_number?
+  after_save :increase_invoice_number, if: :use_sequential_number?
+
+  # Delegates
+
+  delegate :bill_address,
+           :ship_address,
+           :tax_address,
+           :line_items,
+           :display_item_total,
+           :all_adjustments,
+           :shipments,
+           :display_total,
+           :payments, to: :invoiceable
```

Comment:
Many of the delegates will probably turn into methods, and yes: the invoice model will know stuff about its invoiceables. I prefer to have that logic in the model though to having it in the view, because then, the *view* would have to be aware of the inner implementation of the invoiceables - not so nice...